swim/events.go

```diff
@@ -71,6 +71,23 @@ type JoinCompleteEvent struct {
 	Joined    []string      `json:"joined"`
 }
 
+// JoinFailedReason indicates the reason a join failed
+type JoinFailedReason string
+
+const (
+	// ERROR as a JoinFailedReason indicates that the join failed because of an error
```

Comment:
Is there a reason to ALL CAPS these constants? It's inconsistent with other constants that you've defined in this PR. Convention seems to be `PascalCase`